File: sdk/typescript/src/artifact_detect.ts
/**
 * Artifact detection: scans messages for code blocks, JSON objects, and SQL statements.
 * Port of Go memory/artifact_detect.go.
 */

export interface DetectedArtifact {
  content: string;
  artifactType: string; // "code", "json", "sql"
  language: string;
  sourceRole: string; // "user" or "assistant"
}

const codeFenceRe = /```(\w*)\n([\s\S]*?)```/g;
const sqlPrefixRe = /^\s*(SELECT|INSERT|UPDATE|DELETE|CREATE|ALTER|DROP)\b/i;
const jsonBlockRe = /\{[\s\S]*?\}/g;

export function detectArtifacts(
  messages: Array<{ role: string; content: string }>
): DetectedArtifact[] {
  const seen = new Set<string>();
  const result: DetectedArtifact[] = [];

  function add(a: DetectedArtifact): void {
    if (seen.has(a.content)) return;
    seen.add(a.content);
    result.push(a);
  }

  for (const msg of messages) {
    if (!msg) continue;
    const role = msg.role;
    if (role !== 'user' && role !== 'assistant') continue;
    const content = msg.content;

    // 1. Code fences.
    codeFenceRe.lastIndex = 0;
    let match: RegExpExecArray | null;
    while ((match = codeFenceRe.exec(content)) !== null) {
      const lang = (match[1] ?? '').trim();
      const body = (match[2] ?? '').trim();
      if (!body) continue;
      add({
        content: body,
        artifactType: 'code',
        language: lang,
        sourceRole: role,
      });
    }

    // 2. Inline code blocks: consecutive lines starting with 4+ spaces or tab, > 3 lines.
    const stripped = content.replace(codeFenceRe, '');
    const lines = stripped.split('\n');
    let block: string[] = [];
    for (const line of lines) {
      if (line.length > 0 && (line.startsWith('    ') || line[0] === '\t')) {
        block.push(line);
      } else {
        if (block.length > 3) {
          const body = block.join('\n').trim();
          if (body) {
            add({
              content: body,
              artifactType: 'code',
              language: '',
              sourceRole: role,
            });
          }
        }
        block = [];
      }
    }
    if (block.length > 3) {
      const body = block.join('\n').trim();
      if (body) {
        add({
          content: body,
          artifactType: 'code',
          language: '',
          sourceRole: role,
        });
      }
    }

    // 3. JSON objects.
    jsonBlockRe.lastIndex = 0;
    while ((match = jsonBlockRe.exec(stripped)) !== null) {
      const trimmed = match[0].trim();
      try {
        JSON.parse(trimmed);
        add({
          content: trimmed,
          artifactType: 'json',
          language: '',
          sourceRole: role,
        });
      } catch {
        // Not valid JSON — skip.
      }
    }

    // 4. SQL statements.
    for (const line of lines) {
      const trimmed = line.trim();
      if (sqlPrefixRe.test(trimmed)) {
        add({
          content: trimmed,
          artifactType: 'sql',
          language: '',
          sourceRole: role,
        });
      }
    }
  }

  return result;
}

File: sdk/python/memg/artifact_detect.py
"""Artifact detection: scans messages for code blocks, JSON objects, SQL statements."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import List


@dataclass
class DetectedArtifact:
    content: str
    artifact_type: str   # "code", "json", "sql"
    language: str
    source_role: str     # "user" or "assistant"


_CODE_FENCE_RE = re.compile(r"```(\w*)\n([\s\S]*?)```")
_SQL_PREFIX_RE = re.compile(r"(?i)^\s*(SELECT|INSERT|UPDATE|DELETE|CREATE|ALTER|DROP)\b")
_JSON_BLOCK_RE = re.compile(r"(?s)\{[\s\S]*?\}")


def detect_artifacts(messages: List[dict]) -> List[DetectedArtifact]:
    """Scan messages for code blocks, JSON objects, SQL statements.

    messages: list of dicts with 'role' and 'content' keys.
    Returns deduplicated list of DetectedArtifact.
    """
    seen: set = set()
    result: List[DetectedArtifact] = []

    def _add(a: DetectedArtifact) -> None:
        if a.content in seen:
            return
        seen.add(a.content)
        result.append(a)

    for msg in messages:
        if msg is None:
            continue
        role = msg.get("role", "")
        if role not in ("user", "assistant"):
            continue
        content = msg.get("content", "")
        if not content:
            continue

        # 1. Code fences.
        for match in _CODE_FENCE_RE.finditer(content):
            lang = match.group(1).strip()
            body = match.group(2).strip()
            if not body:
                continue
            _add(DetectedArtifact(
                content=body,
                artifact_type="code",
                language=lang,
                source_role=role,
            ))

        # 2. Inline code blocks: consecutive lines starting with 4+ spaces or tab, > 3 lines.
        stripped = _CODE_FENCE_RE.sub("", content)
        lines = stripped.split("\n")
        block: List[str] = []
        for line in lines:
            if len(line) > 0 and (line.startswith("    ") or line[0] == "\t"):
                block.append(line)
            else:
                if len(block) > 3:
                    body = "\n".join(block).strip()
                    if body:
                        _add(DetectedArtifact(
                            content=body,
                            artifact_type="code",
                            language="",
                            source_role=role,
                        ))
                block = []
        if len(block) > 3:
            body = "\n".join(block).strip()
            if body:
                _add(DetectedArtifact(
                    content=body,
                    artifact_type="code",
                    language="",
                    source_role=role,
                ))

        # 3. JSON objects.
        for match in _JSON_BLOCK_RE.finditer(stripped):
            trimmed = match.group(0).strip()
            try:
                json.loads(trimmed)
            except (json.JSONDecodeError, ValueError):
                continue
            _add(DetectedArtifact(
                content=trimmed,
                artifact_type="json",
                language="",
                source_role=role,
            ))

        # 4. SQL statements.
        for line in lines:
            trimmed = line.strip()
            if _SQL_PREFIX_RE.match(trimmed):
                _add(DetectedArtifact(
                    content=trimmed,
                    artifact_type="sql",
                    language="",
                    source_role=role,
                ))

    return result
